lib/Ethereum/Contract.pm

```
@@ -0,0 +1,268 @@
+use Ethereum::Contract::ContractResponse;
+use Ethereum::Contract::ContractTransaction;
+use Ethereum::Contract::Helper::UnitConversion;
+
+has contract_address => ( is => 'rw' );
+has contract_abi     => ( is => 'ro', required => 1 );
+has rpc_client       => ( is => 'ro', default => sub { Ethereum::RPC::Client->new } );
+has defaults         => ( is => 'rw' );
+
+my $contract_decoded = {};
+
+my $meta = __PACKAGE__->meta;
+
+=head2 BUILD
+
+Constructor: Here we get all functions from the passed ABI and bring it to contract class subs.
+
+Parameters: 
+    contract_address (Optional - only if the contract already exists), 
+    contract_abi (Required - https://solidity.readthedocs.io/en/develop/abi-spec.html), 
+    rpc_client (Optional - Ethereum::RPC::Client(https://github.com/binary-com/perl-Ethereum-RPC-Client) - if not given, new instance will be created);
+    defaults (Optional - gas, from)
+    
+Return:
+    New contract instance
+
+=cut
+
+sub BUILD {
+    
+    $meta->make_mutable;
+    
+    my ($self) = @_;
+    
+    my @decoded_json = @{decode_json($self->contract_abi)};
+    my $actions = {};
+    
+    foreach my $json_input (@decoded_json) {
+        if ($json_input->{type}  eq 'function') {
+            
+            my $name = $json_input->{name};
+            my @inputs = @{$json_input->{inputs}};
+            
+            $meta->add_method( $name => sub {
+                
+                my ($self, @params, $payable) = @_;
+                
+                my $function_id = substr($self->get_function_id($name, @inputs), 0, 10);
+                
+                my $res = $self->call($function_id, \@params, $payable);
+                
+                return $res;
+                
+            });
+            
+            $contract_decoded->{$name} = \@inputs;
+            
+        }
+    }
+    
+    $self->defaults->{gas} = Ethereum::Contract::Helper::UnitConversion::to_wei($self->defaults->{gas} // 4000000);
```

The constant `4000000` should not be silently used. Consider something client-interface something like 

```
use  Etherium::Contract qw(default_gas => 4000000)
```

(other defaults can be added here similar way